Add more keyboard mnemonics

Timwi opened the pull request:
## Proposed changes

Hello! I’m back with more keyboard mnemonics:

* All elements in “Remote Repositories” dialog
* The “Fetch and Prune” command in the remote right-click menu
* “Load SSH key” in “Push” dialog

## Screenshots

![master](https://github.com/user-attachments/assets/e8a521cd-4ee7-4fd4-8bb8-fef27e26d7f3)

## Mnemonics added in “Remote Repositories”

```
E	Separate Push Url (checkbox; R tab)
F	Default merge with (textbox/dropdown; D tab)
K	Private key file (textbox; R tab)
L	Load SSH key (button; R tab)
L	Local branch name (textbox; D tab)
N	Name (textbox; R tab)
P	Push Url (textbox; R tab)
R	Remote repository (textbox/dropdown; D tab)
S	Save changes (button)
T	Test connection (button; R tab)
U	Url/Fetch Url (textbox; R tab)
```

## Contribution

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

pmiossec commented on src/app/GitUI/CommandsDialogs/FormRemotes.cs:
```diff
@@ -61,10 +61,10 @@ static SortableGitRefList()
 
         private readonly TranslationString _labelUrlAsFetch =
-            new("Fetch Url");
+            new("Fetch &Url");
```
Is it a good idea to give the same access key for 2 different  labels? What is the behavior ? Shouldn't we use 'f' instead ?

RussKie replied:
Ideally it should be a single mnemonic per control. Ideally, the same operations across the app should use the same mnemonics (e.g., if one dialog uses "&Fetch" and another uses "F&etch" - it is a bad UX).
Also, strictly speaking, not all controls require a mnemonic (as an example look though VS menus; in the MS Office suite not every has mnemonics either, though it's harder to see with the ribbons).

pmiossec commented on src/app/GitUI/CommandsDialogs/FormPush.Designer.cs:
```diff
@@ -72,9 +72,9 @@ private void InitializeComponent()
             SuspendLayout();
-            // 
+            //
             // ForcePushOptionPanel
-            // 
+            //
```
I thought the version without spaces was the newest designer one and so should be kept. Or I'm wrong...

pmiossec replied:
But if it is the case, a PR with only this change on all designer files is maybe a better idea...

RussKie commented:
This doesn't look right:
![image](https://github.com/user-attachments/assets/13e27d66-0029-445d-a877-635f3dd7e18f)

Could you please double check this?

RussKie commented:
@Timwi could you please check the Copliot's feedback. Thank you

Timwi commented:
# Responses

* Fixed the removed spaces.
* “Browse” button: I wasn’t easily able to give different mnemonics to two Browse buttons in the same dialog because they use the same translatable string inside of `FolderBrowserButton`. I remedied this by overriding the `Text` property for that control, allowing me to set it to different strings. This added multiple new “Browse...” strings to the translation file.
* @pmiossec (and Copilot): `&Url` and `Fetch &Url` are never shown at the same time. They refer to conceptually the same item from the user’s perspective (even if they’re separate textbox objects under the hood). Therefore, I think giving them the same mnemonic is preferable. It means that the same mnemonic will always take you to the same place in the dialog regardless of the state it’s in.
* Fixed the unwanted `&` in the listview column header.

## Additionally,

* I noticed that the “Push Url” textbox (when “Separate Push Url” is checked) was still missing a mnemonic, so I added that
* I changed “De&fault merge with” to “&Default merge with”. I initially had it at F because I thought I could put mnemonics on the tabs in the dialog, so I was going to use D for that, but those don’t appear to accept mnemonics. (Which is not a big deal because Ctrl+Tab works fine.)
* The Save Changes button got its mnemonics on just one of the two tabs — added the other one as well.
* Since `FolderBrowserButton` now has a `Text` property, I added a mnemonic for it in the Push and Pull dialogs as well (both W).

New mnemonic assignment for FormRemotes:

```
A	Separate Push Url (checkbox; R tab)
B	Browse (Url/Fetch Url; button; R tab)
D	Default merge with (textbox/dropdown; D tab)
E	Browse (Private key file; button; R tab)
K	Private key file (textbox; R tab)
L	Load SSH key (button; R tab)
N	Name (textbox; R tab)
O	Local branch name (textbox; D tab)
P	Push Url (textbox; R tab)
R	Remote repository (textbox/dropdown; D tab)
S	Save changes (button; both tabs)
T	Test connection (button; R tab)
U	Url/Fetch Url (textbox; R tab)
W	Browse (Push Url; button; R tab)
```

New screenshots (right-click menu not included because no change from original report):

![temp](https://github.com/user-attachments/assets/87c6d122-9ea4-4d7d-a47b-c5fe73e9938a)

pmiossec commented:
>@pmiossec (and Copilot): &Url and Fetch &Url are never shown at the same time. 

I didn't remember how the options were working so that is the answer I wish for: 👍

pmiossec reviewed (commented):
More a discussion for team while we are at that:

I nearly never use this form but if I will and that I use the keyboard I would find much more useful to have mnemonic "&Rebase current branch on top of remote branch, creates linear history (use with caution)" because the merge option is probably more often changed than remote. And so that the merge option will be "&Merge" / "&Rebase" / "&fetch".

What do you think?

Timwi commented on src/app/GitUI/CommandsDialogs/FormRemotes.Designer.cs:
```diff
@@ -668,7 +670,7 @@ private void InitializeComponent()
             label4.Name = "label4";
             label4.Size = new Size(108, 15);
             label4.TabIndex = 0;
-            label4.Text = "Local branch name";
```
Because L was given to “Load SSH key”. I realize that’s on the other tab so it’s never shown at the same time, so it shouldn’t be a problem, but I was just overly cautious. I’m happy to change it if you want me to.

pmiossec replied:
After a check it doesn't interfered, I prefer, yes.

Timwi commented:
> I nearly never use this form but if I will and that I use the keyboard I would find much more useful to have mnemonic "&Rebase current branch on top of remote branch, creates linear history (use with caution)" because the merge option is probably more often changed than remote. And so that the merge option will be "&Merge" / "&Rebase" / "&fetch".

This could certainly be changed, but it’s not in scope for this PR. Those mnemonics were originally added in adda4cb88e66bd4e813462e589fa71cd0378b9ee and aea50a86f441c561864e3d0fc3402777d37c8728 (3 years ago).

Timwi commented:
Worth noting that “rebase” is associated with E in other places:

* Main menu: Commands → R&ebase (R is Reset changes)
* Keyboard shortcut for rebase is Ctrl+Shift+E (Ctrl+Shift+R is show/hide remote branches in list of commits)

However, it’s R in the right-click menu of a commit. Here, E is “Rename branch” (not visible unless selected commit is on a branch).

pmiossec commented:
> Worth noting that “rebase” is associated with E in other places:

Surely a very good point...

mstv commented on src/app/GitUI/UserControls/FolderBrowserButton.cs:
```diff
@@ -18,6 +18,21 @@ public FolderBrowserButton()
+        /// <summary>
+        /// Specifies the text label of the button.
+        /// </summary>
+        public override string Text
+        {
+            get
+            {
+                return buttonBrowse.Text;
+            }
+            set
+            {
+                buttonBrowse.Text = value;
+            }
+        }
```
The following might reduce noise in the translation file:
```suggestion
        public override string _NO_TRANSLATE_Text
        {
            get => buttonBrowse.Text;
            set => buttonBrowse.Text = value;
        }
```

mstv replied:
My main point was and is to add the prefix `_NO_TRANSLATE_`.
Though it cannot be applied here, because it is an overridden property (what I did not notice in the first comment).
But it should work with renaming `buttonBrowse` (and setting the overridden property, too).
```suggestion
        public override string Text
        {
            get => base.Text;
            set
            {
                base.Text = value;
                _NO_TRANSLATE_buttonBrowse.Text = value;
            }
        }
```

mstv replied:
Does my last suggestion work?

mstv replied:
There is no need override `GitExtensionsControl.Text`.
```suggestion
        public string _NO_TRANSLATE_ButtonText
        {
            get => buttonBrowse.Text;
            set => buttonBrowse.Text = value;
        }
```
Maybe, this does not even need that prefix, i.e.
```suggestion
        public string ButtonText
        {
            get => buttonBrowse.Text;
            set => buttonBrowse.Text = value;
        }
```

mstv replied:
Your variant works well, too.

mstv reviewed (commented):
The screenshots LGTM

Timwi commented:
Done. Feel free to squash.

Timwi commented:
Rebased on newest changes.

Timwi commented:
# More progress, yay!

* Rebased on newest changes!
* Changed “L&ocal branch name” to “&Local branch name” as requested — doesn’t conflict with “&Load SSH key” on the other tab
* Regarding the “NO_TRANSLATE”: I understand what you mean by “noise in the translation file” but I’ve looked through them and I found that almost all of them are in dialogs that I’m still likely to fill with mnemonics in the future. Just one of them, `FormFixHome`, I decided to do right here right now!

# New screenshots
![temp](https://github.com/user-attachments/assets/b82fc389-4071-4554-9d64-8226be673125)

mstv commented on src/app/GitUI/Translation/English.xlf:
```diff
@@ -2166,6 +2166,10 @@ Suitable for some config files modified locally.</source>
   </file>
   <file datatype="plaintext" original="FolderBrowserButton" source-language="en">
     <body>
+      <trans-unit id="$this.Text">
+        <source>Browse...</source>
+        <target />
+      </trans-unit>
```
> Regarding the “NO_TRANSLATE”: I understand what you mean by “noise in the translation file” but I’ve looked through them and I found that almost all of them are in dialogs that I’m still likely to fill with mnemonics in the future.

Such additional entries in the translation file are wrong because you have not added any new control.

Timwi replied:
You’re absolutely right, I understand what you’re saying now. The changes have been made.

mstv reviewed (commented):
> New screenshots

Rather udpate the PR description, please. (It even has a history.)

Timwi commented:
> Rather udpate the PR description, please. (It even has a history.)

Sure thing — done!

mstv reviewed (approved):
Thank you

RussKie commented on src/app/GitUI/CommandsDialogs/FormRemotes.Designer.cs:
```diff
@@ -291,6 +291,7 @@ private void InitializeComponent()
             folderBrowserButtonUrl.PathShowingControl = Url;
             folderBrowserButtonUrl.Size = new Size(104, 25);
             folderBrowserButtonUrl.TabIndex = 2;
+            folderBrowserButtonUrl.Text = "&Browse...";
```
[nit] This looks inconsistent with other "Browse" buttons 

We may need to review and make this consistency across the app (if possible)

mstv replied:
The default mnemonic could be "W", and this default could by assigned in `FolderBrowserButton`.

Though, there is at least one form with two browse buttons. I.e. consistency of mnemonics has limits. IMO the priority is to *have* mnemonics. It does not confuse me if they are different in different dialogs, because they are underlined (with the appropriate Windows setting - unfortunately, only while `Alt` is pressed otherwise).

![image](https://github.com/user-attachments/assets/0e0e4024-2963-428f-a74d-f857b9a84060)

RussKie replied:
Some dialogs may have several "browse" buttons (like in this case), but rarely (if at all) those buttons have the same behaviour.
UI consistency is very important, IMO, as users learn that a specific shortcut performs a specific operation. Imagine, ctrl+c or alt+f4 behaved differently in different dialogs.

gerhardol replied:
+1 for me, agree that this should be consistent.

Timwi replied:
Done!

RussKie commented on src/app/GitUI/CommandsDialogs/SettingsDialog/Pages/FormFixHome.Designer.cs:
```diff
@@ -66,6 +66,7 @@ private void InitializeComponent()
             otherHomeBrowse.PathShowingControl = otherHomeDir;
             otherHomeBrowse.Size = new Size(130, 25);
             otherHomeBrowse.TabIndex = 5;
+            otherHomeBrowse.Text = "&Browse...";
```
Ditto

Timwi replied:
Done!

RussKie reviewed (commented):
Reviewed on a phone

Timwi commented:
I have done as requested and made `"&Browse..."` the default text for the FolderBrowserButton.

Furthermore, I’ve changed the following things:

* ForkAndCloneForm: replaced the plain Browse button with a FolderBrowserButton
* FormClone: changed `"&Branch:"` to `"Br&anch:"` to resolve the conflict with the Browse button (the conflict was already there)
* FormPull: Removed `"Bro&wse..."` causing the mnemonic to change to the default of B. According to the list of mnemonics in the comment at the top of the file, B was available. (Curiously, it says that W is assigned to something, but that something doesn’t show up in the dialog for me, but maybe it’s contextual.)
* FormPush: changed `"&Branch to push"` to `"Br&anch to push"` to enable the Browse button to get the default `B` mnemonic.

mstv reviewed (commented):
> I have done as requested and made `"&Browse..."` the default text for the FolderBrowserButton.

🤔 Unfortunately, this has the side-effect that the default text is not translated any longer (refer to https://github.com/gitextensions/gitextensions/commit/3ead3527ff8c77357e9dc02b96c45524076d5935#diff-e4e50022b516b7c5dfa785a55a5df8569f48403871359b2b33cbc577451e34e5).
Possible workaround: Use a `TranslationString` as the default label in `FolderBrowserButton`.

> Furthermore, I’ve changed the following things:
> 
> * ForkAndCloneForm: replaced the plain Browse button with a FolderBrowserButton

👍

> * FormClone: changed `"&Branch:"` to `"Br&anch:"` to resolve the conflict with the Browse button (the conflict was already there)

OK for me.
 
> * FormPull: Removed `"Bro&wse..."` causing the mnemonic to change to the default of B. According to the list of mnemonics in the comment at the top of the file, B was available.

👍

> (Curiously, it says that W is assigned to something, but that something doesn’t show up in the dialog for me, but maybe it’s contextual.)

ℹ️ only if cloned without full history: https://github.com/gitextensions/gitextensions/blob/aa51216b5ed47fa79dea8d2c93c3a3614136d90a/src/app/GitUI/CommandsDialogs/FormPull.cs#L204-L210

> * FormPush: changed `"&Branch to push"` to `"Br&anch to push"` to enable the Browse button to get the default `B` mnemonic.

❗ In this form, "&Branch" is an important mnemonic. It takes precedence over "&Browse", which is even disabled by default.

Timwi commented:
> 🤔 Unfortunately, this has the side-effect that the default text is not translated any longer

Well, this is mostly because of https://github.com/gitextensions/gitextensions/commit/89b0f93adaa6a766a6293ab041311df76352fad7 where we added the NO_TRANSLATE; but we also changed the `Text` property to use a `base` call. I changed it back so it’s no longer NO_TRANSLATE but I kept the `base` call. This appears to have the effect that we want: there’s now one translatable string for the default value and one for every overridden variant.

> ❗ In this form, "&Branch" is an important mnemonic. It takes precedence over "&Browse", which is even disabled by default.

Gotcha. Changed Branch to B and Browse to W.

mstv reviewed (approved):
Thank you!
